File: crates/tacet-napi/ts/result.ts
/**
 * Result wrapper class with helper methods.
 */

import {
  type AnalysisResult,
  type EffectEstimate,
  type Diagnostics,
  Outcome,
  type InconclusiveReason,
  type Exploitability,
  type MeasurementQuality,
} from "../index.js";

import type { BatchingInfo, TimerInfo } from "./collector.js";
import { TimingLeakError } from "./errors.js";

// Reverse mappings for const enums (they don't have runtime reverse mappings)
const OUTCOME_NAMES = ["Pass", "Fail", "Inconclusive", "Unmeasurable"] as const;
const EXPLOITABILITY_NAMES = [
  "SharedHardwareOnly",
  "Http2Multiplexing",
  "StandardRemote",
  "ObviousLeak",
] as const;
const QUALITY_NAMES = ["Excellent", "Good", "Poor", "TooNoisy"] as const;
const INCONCLUSIVE_NAMES = [
  "None",
  "DataTooNoisy",
  "NotLearning",
  "WouldTakeTooLong",
  "TimeBudgetExceeded",
  "SampleBudgetExceeded",
  "ConditionsChanged",
  "ThresholdElevated",
] as const;

/**
 * Wrapper around AnalysisResult with helper methods for ergonomic result handling.
 *
 * @example
 * ```typescript
 * const result = TimingOracle
 *   .forAttacker(AttackerModel.AdjacentNetwork)
 *   .test(inputs, operation);
 *
 * // Check outcome
 * if (result.isFail()) {
 *   console.error(`Leak: ${result.exploitabilityString()}`);
 * }
 *
 * // Or assert directly
 * result.assertNoLeak();  // throws TimingLeakError if Fail
 *
 * // Or log summary
 * console.log(result.toString());
 * ```
 */
export class TimingTestResult {
  constructor(
    private readonly raw: AnalysisResult,
    /** Batching configuration from pilot phase. */
    readonly batchingInfo: BatchingInfo,
    /** Timer calibration info. */
    readonly timerInfo: TimerInfo,
  ) {}

  // Forward all AnalysisResult properties as getters

  /** Test outcome. */
  get outcome(): Outcome {
    return this.raw.outcome;
  }

  /** Leak probability: P(max_k |(X*beta)_k| > theta | data). */
  get leakProbability(): number {
    return this.raw.leakProbability;
  }

  /** Effect size estimate. */
  get effect(): EffectEstimate {
    return this.raw.effect;
  }

  /** Measurement quality. */
  get quality(): MeasurementQuality {
    return this.raw.quality;
  }

  /** Number of samples used per class. */
  get samplesUsed(): number {
    return this.raw.samplesUsed;
  }

  /** Time spent in seconds. */
  get elapsedSecs(): number {
    return this.raw.elapsedSecs;
  }

  /** Exploitability (only meaningful if outcome == Fail). */
  get exploitability(): Exploitability {
    return this.raw.exploitability;
  }

  /** Inconclusive reason (only meaningful if outcome == Inconclusive). */
  get inconclusiveReason(): InconclusiveReason {
    return this.raw.inconclusiveReason;
  }

  /** Minimum detectable effect in nanoseconds. */
  get mdeNs(): number {
    return this.raw.mdeNs;
  }

  /** Timer resolution in nanoseconds. */
  get timerResolutionNs(): number {
    return this.raw.timerResolutionNs;
  }

  /** User's requested threshold (theta) in nanoseconds. */
  get thetaUserNs(): number {
    return this.raw.thetaUserNs;
  }

  /** Effective threshold after floor adjustment in nanoseconds. */
  get thetaEffNs(): number {
    return this.raw.thetaEffNs;
  }

  /** Recommendation string (empty if not applicable). */
  get recommendation(): string {
    return this.raw.recommendation;
  }

  /** Detailed diagnostics. */
  get diagnostics(): Diagnostics {
    return this.raw.diagnostics;
  }

  // Predicate methods

  /** True if outcome is Pass. */
  isPass(): boolean {
    return this.outcome === Outcome.Pass;
  }

  /** True if outcome is Fail. */
  isFail(): boolean {
    return this.outcome === Outcome.Fail;
  }

  /** True if outcome is Inconclusive. */
  isInconclusive(): boolean {
    return this.outcome === Outcome.Inconclusive;
  }

  /** True if outcome is Unmeasurable. */
  isUnmeasurable(): boolean {
    return this.outcome === Outcome.Unmeasurable;
  }

  /** True if outcome is Pass or Fail (a decision was reached). */
  isConclusive(): boolean {
    return this.isPass() || this.isFail();
  }

  // String formatters (enums don't have reverse mappings in the napi-rs output)

  /** Get outcome as a string: 'Pass', 'Fail', 'Inconclusive', or 'Unmeasurable'. */
  outcomeString(): string {
    return OUTCOME_NAMES[this.outcome];
  }

  /** Get exploitability as a string. */
  exploitabilityString(): string {
    return EXPLOITABILITY_NAMES[this.exploitability];
  }

  /** Get measurement quality as a string. */
  qualityString(): string {
    return QUALITY_NAMES[this.quality];
  }

  /** Get inconclusive reason as a string. */
  inconclusiveReasonString(): string {
    return INCONCLUSIVE_NAMES[this.inconclusiveReason];
  }

  // Helper methods

  /** Format leak probability as a percentage string (e.g., "12.3%"). */
  leakProbabilityPercent(): string {
    return `${(this.leakProbability * 100).toFixed(1)}%`;
  }

  /** Total effect magnitude in nanoseconds. */
  totalEffectNs(): number {
    return this.effect.maxEffectNs;
  }

  /**
   * Assert that no timing leak was detected.
   *
   * Throws `TimingLeakError` if outcome is Fail.
   * Returns `this` for chaining if outcome is not Fail.
   *
   * @example
   * ```typescript
   * result.assertNoLeak();  // throws if Fail
   * ```
   */
  assertNoLeak(): this {
    if (this.isFail()) {
      throw new TimingLeakError(this);
    }
    return this;
  }

  /**
   * Get a human-readable summary string.
   *
   * @example
   * ```typescript
   * console.log(result.toString());
   * // → "Pass: P(leak)=1.2%"
   * // → "Fail: P(leak)=99.8%, effect=142.3ns (Http2Multiplexing)"
   * // → "Inconclusive: P(leak)=45.2% (DataTooNoisy)"
   * ```
   */
  toString(): string {
    let msg = `${this.outcomeString()}: P(leak)=${this.leakProbabilityPercent()}`;
    if (this.isFail()) {
      msg += `, effect=${this.totalEffectNs().toFixed(1)}ns (${this.exploitabilityString()})`;
    }
    if (this.isInconclusive()) {
      msg += ` (${this.inconclusiveReasonString()})`;
    }
    return msg;
  }
}

File: crates/tacet-go/result.go
package tacet

import (
	"fmt"
	"time"
)

// Outcome represents the test result.
type Outcome int

const (
	// Pass indicates no timing leak was detected within the threshold.
	Pass Outcome = iota
	// Fail indicates a timing leak was detected exceeding the threshold.
	Fail
	// Inconclusive indicates the test could not reach a decision.
	Inconclusive
	// Unmeasurable indicates the operation is too fast to measure reliably.
	Unmeasurable
)

// String returns the string representation of the outcome.
func (o Outcome) String() string {
	switch o {
	case Pass:
		return "Pass"
	case Fail:
		return "Fail"
	case Inconclusive:
		return "Inconclusive"
	case Unmeasurable:
		return "Unmeasurable"
	default:
		return "Unknown"
	}
}

// InconclusiveReason explains why a test was inconclusive.
type InconclusiveReason int

const (
	ReasonNone InconclusiveReason = iota
	ReasonDataTooNoisy
	ReasonNotLearning
	ReasonWouldTakeTooLong
	ReasonTimeBudgetExceeded
	ReasonSampleBudgetExceeded
	ReasonConditionsChanged
	ReasonThresholdElevated
	ReasonModelMismatch
)

// String returns the string representation of the reason.
func (r InconclusiveReason) String() string {
	switch r {
	case ReasonNone:
		return ""
	case ReasonDataTooNoisy:
		return "DataTooNoisy"
	case ReasonNotLearning:
		return "NotLearning"
	case ReasonWouldTakeTooLong:
		return "WouldTakeTooLong"
	case ReasonTimeBudgetExceeded:
		return "TimeBudgetExceeded"
	case ReasonSampleBudgetExceeded:
		return "SampleBudgetExceeded"
	case ReasonConditionsChanged:
		return "ConditionsChanged"
	case ReasonThresholdElevated:
		return "ThresholdElevated"
	case ReasonModelMismatch:
		return "ModelMismatch"
	default:
		return "Unknown"
	}
}

// Exploitability assesses the practical exploitability of a detected leak.
type Exploitability int

const (
	// SharedHardwareOnly: < 10 ns - requires shared hardware (SGX, containers) to exploit.
	SharedHardwareOnly Exploitability = iota
	// HTTP2Multiplexing: 10-100 ns - exploitable via HTTP/2 request multiplexing.
	HTTP2Multiplexing
	// StandardRemote: 100 ns - 10 us - exploitable with standard remote timing.
	StandardRemote
	// ObviousLeak: > 10 us - obvious leak, trivially exploitable.
	ObviousLeak
)

// String returns the string representation of exploitability.
func (e Exploitability) String() string {
	switch e {
	case SharedHardwareOnly:
		return "SharedHardwareOnly"
	case HTTP2Multiplexing:
		return "HTTP2Multiplexing"
	case StandardRemote:
		return "StandardRemote"
	case ObviousLeak:
		return "ObviousLeak"
	default:
		return "Unknown"
	}
}

// Quality assesses the measurement quality.
type Quality int

const (
	// Excellent: MDE < 5 ns - excellent measurement precision.
	Excellent Quality = iota
	// Good: MDE 5-20 ns - good precision for most use cases.
	Good
	// Poor: MDE 20-100 ns - limited precision.
	Poor
	// TooNoisy: MDE > 100 ns - too noisy for reliable detection.
	TooNoisy
)

// String returns the string representation of quality.
func (q Quality) String() string {
	switch q {
	case Excellent:
		return "Excellent"
	case Good:
		return "Good"
	case Poor:
		return "Poor"
	case TooNoisy:
		return "TooNoisy"
	default:
		return "Unknown"
	}
}

// Effect holds the effect size estimate.
type Effect struct {
	// MaxEffectNs is the maximum effect in nanoseconds: max_k |delta_k|.
	MaxEffectNs float64
	// CredibleIntervalNs is the 95% credible interval (lower, upper).
	CredibleIntervalNs [2]float64
	// TopQuantiles contains the top 2-3 quantiles by exceedance probability.
	TopQuantiles []TopQuantile
}

// TopQuantile represents a quantile with exceedance information.
type TopQuantile struct {
	// QuantileP is the quantile level (e.g., 0.9 for 90th percentile).
	QuantileP float64
	// MeanNs is the posterior mean delta_k in nanoseconds.
	MeanNs float64
	// CI95Ns is the 95% marginal credible interval (lower, upper).
	CI95Ns [2]float64
	// ExceedProb is P(|delta_k| > theta_eff | data).
	ExceedProb float64
}

// Diagnostics holds detailed diagnostic information from the analysis.
// This provides insight into the statistical analysis quality and can
// be used for debugging or understanding measurement reliability.
type Diagnostics struct {
	// DependenceLength is the block length used for bootstrap resampling.
	DependenceLength uint64
	// EffectiveSampleSize accounts for autocorrelation in timing data.
	EffectiveSampleSize uint64
	// StationarityRatio is the ratio of post-test variance to calibration variance.
	StationarityRatio float64
	// StationarityOK indicates whether the stationarity check passed.
	StationarityOK bool

	// DiscreteMode indicates whether discrete mode was used (low timer resolution).
	DiscreteMode bool
	// TimerResolutionNs is the timer resolution in nanoseconds.
	TimerResolutionNs float64

	// LambdaMean is the posterior mean of the latent scale parameter lambda.
	LambdaMean float64
	// LambdaSD is the posterior standard deviation of lambda.
	LambdaSD float64
	// LambdaESS is the effective sample size of the lambda chain.
	LambdaESS float64
	// LambdaMixingOK indicates whether the lambda chain mixed well.
	LambdaMixingOK bool

	// KappaMean is the posterior mean of the likelihood precision kappa.
	KappaMean float64
	// KappaCV is the coefficient of variation of kappa.
	KappaCV float64
	// KappaESS is the effective sample size of the kappa chain.
	KappaESS float64
	// KappaMixingOK indicates whether the kappa chain mixed well.
	KappaMixingOK bool
}


// Result holds the complete analysis result.
type Result struct {
	// Outcome is the test result (Pass, Fail, Inconclusive, or Unmeasurable).
	Outcome Outcome

	// LeakProbability is P(max_k |(X*beta)_k| > theta | data).
	// For Pass: typically < 5%. For Fail: typically > 95%.
	LeakProbability float64

	// Effect is the estimated timing effect.
	Effect Effect

	// Quality is the measurement quality assessment.
	Quality Quality

	// SamplesUsed is the number of samples collected per class.
	SamplesUsed uint64

	// ElapsedTime is how long the test took.
	ElapsedTime time.Duration

	// Exploitability assesses practical exploitability (only meaningful for Fail).
	Exploitability Exploitability

	// InconclusiveReason explains why the test was inconclusive (if applicable).
	InconclusiveReason InconclusiveReason

	// MDENs is the minimum detectable effect in nanoseconds.
	MDENs float64

	// TimerResolutionNs is the timer resolution in nanoseconds.
	TimerResolutionNs float64

	// ThetaUserNs is the user's requested threshold in nanoseconds.
	ThetaUserNs float64

	// ThetaEffNs is the effective threshold after floor adjustment.
	ThetaEffNs float64

	// ThetaFloorNs is the measurement floor (minimum detectable effect given noise).
	ThetaFloorNs float64

	// DecisionThresholdNs is the threshold at which the decision was made.
	DecisionThresholdNs float64

	// Recommendation is guidance for inconclusive/unmeasurable results.
	Recommendation string

	// Diagnostics contains detailed diagnostic information (nil if not available).
	Diagnostics *Diagnostics
}

// IsConclusive returns true if the result is Pass or Fail.
func (r *Result) IsConclusive() bool {
	return r.Outcome == Pass || r.Outcome == Fail
}

// IsMeasurable returns true if the operation was measurable.
func (r *Result) IsMeasurable() bool {
	return r.Outcome != Unmeasurable
}

// String returns a human-readable summary of the result.
func (r *Result) String() string {
	switch r.Outcome {
	case Pass:
		return fmt.Sprintf("Pass: P(leak)=%.1f%%, max_effect=%.2fns, quality=%s, samples=%d",
			r.LeakProbability*100, r.Effect.MaxEffectNs, r.Quality, r.SamplesUsed)
	case Fail:
		return fmt.Sprintf("FAIL: P(leak)=%.1f%%, max_effect=%.2fns, exploitability=%s, samples=%d",
			r.LeakProbability*100, r.Effect.MaxEffectNs, r.Exploitability, r.SamplesUsed)
	case Inconclusive:
		return fmt.Sprintf("Inconclusive (%s): P(leak)=%.1f%%, samples=%d",
			r.InconclusiveReason, r.LeakProbability*100, r.SamplesUsed)
	case Unmeasurable:
		return fmt.Sprintf("Unmeasurable: %s", r.Recommendation)
	default:
		return "Unknown result"
	}
}
